File: src/website/actions/users/getUsers.ts
import { Dispatch } from "react";

import { Option, Timestamp } from "../../domain/model/common";
import { UserId, UserSummary } from "../../domain/model/users";
import { GetUsersRequest } from "../../services/userMgmt/getUsers";
import userMgmtService from "../../services/userMgmt/service";
import { HttpError } from "../../errors/httpError";

export const GET_USERS_REQUESTED = "GET_USERS_REQUESTED";
export const GET_USERS_SUCCEEDED = "GET_USERS_SUCCEEDED";
export const GET_USERS_FAILED = "GET_USERS_FAILED";

export default function(users: UserId[], updatedSince: Option<Timestamp> = null) {
    return async (dispatch: Dispatch<any>) => {
        const request: GetUsersRequest = {
            users,
            updatedSince
        };

        const requestAction: GetUsersRequestedEvent = {
            type: GET_USERS_REQUESTED,
            payload: request
        };

        dispatch(requestAction);

        const result = await userMgmtService.getUsers(request);

        let outcomeEvent;
        if (result.kind === "success") {
            outcomeEvent = {
                type: GET_USERS_SUCCEEDED,
                payload: {
                    request,
                    result: {
                        users: result.users,
                        timestamp: result.timestamp
                    }
                }
            } as GetUsersSucceededEvent;
        } else {
            outcomeEvent = {
                type: GET_USERS_FAILED,
                payload: request,
                httpError: result.kind === "httpError" ? result : undefined
            } as GetUsersFailedEvent;
        }

        dispatch(outcomeEvent);
    }
}

export type GetUsersRequestedEvent = {
    type: typeof GET_USERS_REQUESTED,
    payload: GetUsersRequest
}

export type GetUsersSucceededEvent = {
    type: typeof GET_USERS_SUCCEEDED,
    payload: {
        request: GetUsersRequest,
        result: {
            users: UserSummary[],
            timestamp: Timestamp
        }
    }
}

export type GetUsersFailedEvent = {
    type: typeof GET_USERS_FAILED,
    payload: GetUsersRequest,
    httpError?: HttpError
}

File: src/services/user_mgmt/src/queries/get_users.rs
use self::Response::*;
use crate::domain::user_store::{UserStore, UserSummary};
use ic_cdk::export::candid::CandidType;
use ic_cdk::storage;
use serde::Deserialize;
use shared::user_id::UserId;
use shared::{timestamp, timestamp::Timestamp};

pub fn query(request: Request) -> Response {
    let me = shared::user_id::get_current();
    let user_store: &UserStore = storage::get();
    let now = timestamp::now();

    let users = user_store.get_users(request.users, &me, request.updated_since, now);

    let result = Result {
        users,
        timestamp: now,
    };

    Success(result)
}

#[derive(Deserialize)]
pub struct Request {
    users: Vec<UserId>,
    updated_since: Option<Timestamp>,
}

#[derive(CandidType)]
pub enum Response {
    Success(Result),
}

#[derive(CandidType)]
pub struct Result {
    users: Vec<UserSummary>,
    timestamp: Timestamp,
}
